bootstrap/bootstrap.go

```diff
@@ -293,11 +293,56 @@ func (b *Bootstrap) executeHook(ctx context.Context, hookCfg HookConfig) error {
 
 	b.shell.Headerf("Running %s hook", hookName)
 
+	isBinary, err := shellscript.IsBinaryExecutable(hookCfg.Path)
+	if err != nil {
+		return fmt.Errorf("checking if %s hook is a binary: %w", hookName, err)
+	}
+
+	sb, err := shellscript.ShebangLine(hookCfg.Path)
+	if err != nil {
+		return fmt.Errorf("reading shebang line for %s hook: %w", hookName, err)
+	}
+
+	switch {
+	case isBinary:
+		// It's a binary, so we'll just run it directly, no wrapping needed or possible
+		return b.runUnwrappedHook(ctx, hookName, hookCfg)
+	case shellscript.IsPOSIXShell(sb):
+		// It's definitely a shell script, wrap it so that we can snaffle the changed environment variables
+		return b.runWrappedShellScriptHook(ctx, hookName, hookCfg)
+	case sb != "" && !shellscript.IsPOSIXShell(sb):
+		// It's an interpretable-something with a shebang line, and it almost certainly isn't a shell script,
+		// so it won't be capable of setting environment variables (outside of the job api), so just run it directly
+		return b.runUnwrappedHook(ctx, hookName, hookCfg)
+	case sb == "":
+		// If there's no shebang line in the hook, and it's not a binary, so let's just assume that it's a shell script
+		// The OS will complain if we do something stupid, so we'll just give it a shot
+		return b.runWrappedShellScriptHook(ctx, hookName, hookCfg)
+	default:
+		return fmt.Errorf(`the buildkite agent wasn't able to determine the type of hook it was trying to run.
```

Comment:
I'm interested in (but not opposed to) the reasoning behind magic-number-detecting known executable binary formats, compared to just trusting the `+x` file mode to decide if it's executable, and checking for known-shell shebang to do the env wrapping.

Is it because we support wrapped shell hooks without a shebang?

But in this implementation, it'll even try to run binary-but-not-known-executable files as wrapped shell scripts. I wonder if that's a problem.

I wonder if this could work:

- if `isFileModeExecutable(path) == false` then `return`
- if `IsPOSIXShell(shebang)` then `runWrappedShellScriptHook`
- else `runUnwrappedHook`

That would delegate more trust/control to the `+x` file mode, and mean we don't need to maintain a (beautiful 💚) definition of known executable magic numbers.

There's quite likely some factors I'm not considering.

Reply:
Hmm… I guess a bit difference there is that we'd no longer try running non-shell-shebanged executables as _wrapped_ hooks, which would perhaps break lots of things, if there's non-shebanged shell hooks out in the wild. And I'm sure there it.

So perhaps what I'm suggesting is

- case `IsPOSIXShell(shebang)`: `runWrappedShellScriptHook`
- case `hasShebang || isBinary`: `runUnwrappedHook`
- default (case `isValidUTF8`): `runWrappedShellScriptHook`

Note that in this version, `isBinary` means “isn't valid UTF-8 (and thus also isn't ASCII)” which implies “more likely a compiled program than a shell script”, and there's no magic number detection.

Just a thought 🤷‍♂️

Reply:
I suspect checking for the `x` bit won't work on Windows. But that does not mean we should not respect it on unicies.

Reply:
> ... we'd no longer try running non-shell-shebanged executables as wrapped hooks, which would perhaps break lots of things, if there's non-shebanged shell hooks out in the wild. And I'm sure there it.

yep, this was a big part of the motivation here. in fact, on windows, the only way to identify a batch/powershell script is the file extension - they don't use shebangs over in windowsland, so maintaining the default of "it's not shebanged, i don't know what it is" being "assume it's a shellscript" is sensible in this case IMO.

---

re: alternate ways of detecting executable binaries:

i agree that taking care of a big list of executable magic numbers is a pain in the butt, but i can't think of a better cross-platform way to check if something is an executable - checking if the file encoding is funky and there's an `x` permission bit set is fine on unices, but as @triarius mentions, it probably won't work on windows.

additionally, checking that a file is invalid UTF-8 is necessary, but not sufficient in proving that it's a binary. to bring up windows again, the default encoding for Powershell scripts is Windows-1252, which would fail this check and cause the agent to try to run the powershell scripts unwrapped, which would be a problem.

i guess my point here is that imo, having a very narrow definition of what a binary hook looks like (it must have the right magic number) is a benefit to us; it allows us to define a strict API on what will be executed as a binary hook. if it isn't a valid executable for your platform, we'll try to run it in a different way and it'll break. 

if we've missed binary types in our big list, then we can add their signatures, but were we to use a less exact way of identifying binaries, it would open us up to all sorts of horrible bugs, encoding and otherwise

Reply:
Makes sense! 👍🏼